prov/efa/src/rdm/efa_rdm_error.h

```diff
@@ -0,0 +1,105 @@
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef EFA_RDM_ERROR_H
+#define EFA_RDM_ERROR_H
+
+#include <ctype.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "efa_errno.h"
+#include "efa_rdm_peer.h"
+
+#define HOST_ID_STR_LENGTH 19
+#define HOST_ID_STR_FORMAT "i-%017lx"
+#define NA_STR "N/A"
+#define ENRICHED_ERR_FORMAT "%s My EFA addr: %s My host id: %s Peer EFA addr: %s Peer host id: %s"
+#define CONN_INFO_MAX_LENGTH (sizeof(ENRICHED_ERR_FORMAT) + HOST_ID_STR_LENGTH * 2 + OFI_ADDRSTRLEN * 2)
+
+/**
+ * @brief Write the error message and return its byte length
+ * @param[in]    ep          RXR endpoint
+ * @param[in]    addr        Remote peer fi_addr_t
+ * @param[in]    err         FI_* error code(must be positive)
+ * @param[in]    prov_errno  EFA provider * error code(must be positive)
+ * @param[out]   buf         Pointer to the address of error data written by this function
+ * @param[out]   buflen      Pointer to the returned error data size
+ * @return       A status code. 0 if the error data was written successfully, otherwise a negative FI error code.
+ */
+static inline int efa_rdm_error_write_msg(struct rxr_ep *ep, fi_addr_t addr, int err, int prov_errno, void **buf, size_t *buflen)
+{
+    char ep_addr_str[OFI_ADDRSTRLEN], peer_addr_str[OFI_ADDRSTRLEN];
+    char local_host_id_str[HOST_ID_STR_LENGTH + 1], peer_host_id_str[HOST_ID_STR_LENGTH + 1];
+    const char *base_msg = efa_strerror(prov_errno, NULL);
+    size_t len = 0;
+    struct efa_rdm_peer *peer = rxr_ep_get_peer(ep, addr);
+
+    len = sizeof(ep_addr_str);
+    rxr_ep_raw_addr_str(ep, ep_addr_str, &len);
+    len = sizeof(peer_addr_str);
+    rxr_ep_get_peer_raw_addr_str(ep, addr, peer_addr_str, &len);
+
+    if (prov_errno == FI_EFA_REMOTE_ERROR_RNR) {
+        /* RNR error is retried. No need to write error data */
+        return -FI_EAGAIN;
+    }
+
+    len = strlen(base_msg) + CONN_INFO_MAX_LENGTH + 1;
+
+    memset(local_host_id_str, 0, sizeof(local_host_id_str));
+    if (!ep->host_id || HOST_ID_STR_LENGTH != snprintf(local_host_id_str, HOST_ID_STR_LENGTH + 1, HOST_ID_STR_FORMAT, ep->host_id)) {
+        strcpy(local_host_id_str, NA_STR);
+    }
+
+    memset(peer_host_id_str, 0, sizeof(peer_host_id_str));
+    if (!ep->host_id || HOST_ID_STR_LENGTH != snprintf(peer_host_id_str, HOST_ID_STR_LENGTH + 1, HOST_ID_STR_FORMAT, peer->host_id)) {
+        strcpy(peer_host_id_str, NA_STR);
+    }
+
+    int ret = snprintf(ep->err_msg, len, ENRICHED_ERR_FORMAT, base_msg, ep_addr_str, local_host_id_str, peer_addr_str, peer_host_id_str);
+
+    if (ret < 0 || ret >= len) {
+        return -FI_EINVAL;
+    }
+
+    if (strlen(ep->err_msg) >= RXR_ERROR_MSG_BUFFER_LENGTH) {
+        return -FI_ENOBUFS;
+    }
```

Comment:
Should we set buf to ep->err_msg or maybe to a static "Problem creating enriched error string" or something when returning early?

looking at the strlen(ep->err_msg) check made me realize if that's the relevant buffer size, it should probably be the argument to snprintf on line 90.